File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/repository/SubscriptionRepository.kt
package com.storysphere.storyreader.repository

import com.storysphere.storyreader.model.Subscription
import com.storysphere.storyreader.model.SubscriptionPlan
import com.storysphere.storyreader.network.GraphQLService
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SubscriptionRepository @Inject constructor(
    private val graphQLService: GraphQLService
) {
    suspend fun getSubscription(userId: String): Result<Subscription?> {
        return graphQLService.getSubscription(userId)
    }
    
    suspend fun getSubscriptionPlans(): Result<List<SubscriptionPlan>> {
        return graphQLService.getSubscriptionPlans()
    }
    
    suspend fun subscribe(userId: String, planId: String): Result<Subscription> {
        return graphQLService.subscribe(userId, planId)
    }
    
    suspend fun cancelSubscription(userId: String): Result<Unit> {
        return graphQLService.cancelSubscription(userId)
    }
}

File: Novels/packages/5-mobile-ios/StoryReader/Repository/SubscriptionRepository.swift
import Foundation
import Combine

// Subscription Repository - Subscription logic
class SubscriptionRepository {
    private let offlineService = OfflineService()
    private let graphQLService = GraphQLService.shared
    
    func getSubscription(userId: String) -> AnyPublisher<Subscription?, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    let subscription = try await self.graphQLService.getSubscription(userId: userId)
                    // Cache in Core Data
                    if let subscription = subscription {
                        try? self.offlineService.saveSubscription(subscription)
                    }
                    promise(.success(subscription))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    func getPlans() -> AnyPublisher<[MembershipPlan], Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    let plans = try await self.graphQLService.getSubscriptionPlans()
                    promise(.success(plans))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    func subscribe(userId: String, planId: String) -> AnyPublisher<Subscription, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    let subscription = try await self.graphQLService.subscribe(userId: userId, planId: planId)
                    // Save to Core Data
                    try? self.offlineService.saveSubscription(subscription)
                    promise(.success(subscription))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    func cancelSubscription(userId: String) -> AnyPublisher<Void, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    try await self.graphQLService.cancelSubscription(userId: userId)
                    promise(.success(()))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
}


